File: dart/state/state.dart
/**
 * USANDO O PADRÃO DE PROJECTO STATE
 * Definindo a saida do som em reprodutores de musica, com diferetes estados de estilo
 */

abstract class SoundBassTypeState {
  void soundPlay() {}
}

class RockBaseType implements SoundBassTypeState {
  @override
  void soundPlay() {
    print("Playing [Rock] base type");
  }
}

class JazBaseType implements SoundBassTypeState {
  @override
  void soundPlay() {
    print("Playing [Jazz] base type");
  }
}

class RapBaseType implements SoundBassTypeState {
  @override
  void soundPlay() {
    print("Playing [Rap] base type");
  }
}

class EletronicBaseType implements SoundBassTypeState {
  @override
  void soundPlay() {
    print("Playing [Eletronic] base type");
  }
}

class MusicPlayer {
  SoundBassTypeState soundBassTypeState;

  MusicPlayer(this.soundBassTypeState);

  playSound(){
    this.soundBassTypeState.soundPlay();
  }

  setBasetype(SoundBassTypeState soundBassTypeState){
    this.soundBassTypeState = soundBassTypeState;
  }
}

void main(){

  //Criando o music player e definindo os estados 
  MusicPlayer musicPlayer = MusicPlayer(JazBaseType());
  musicPlayer.playSound();
  //Trocando de estildo de base
  musicPlayer.setBasetype(RapBaseType());
  musicPlayer.playSound();
  
}

File: python/state/state.py
from abc import ABC, abstractmethod

#Padrão STATE
#Criar estadosde tipo de base a ser reproduzido em um leitor de musica, RAP ou Jazz

class SoundBassTypeState(ABC):

    @abstractmethod
    def play_sound():
        pass
        
class RapBassType(SoundBassTypeState):

    def play_sound():
        print("It's playing in [RAP] bass")

class JazzBassType(SoundBassTypeState):

    def play_sound():
         print("It's playing in [JAZZ] bass")

class MusicPlayer():

    sound_bass_type_state: SoundBassTypeState = None

    def __init__(self, sound_bass_type_state: SoundBassTypeState):
        self.sound_bass_type_state = sound_bass_type_state

    def play(self):
        self.sound_bass_type_state.play_sound()

    def set_base_type(self, sound_bass_type_state: SoundBassTypeState):
        self.sound_bass_type_state = sound_bass_type_state



music_player = MusicPlayer(JazzBassType())

music_player.play()
dc

class A(ABC):
    def __init__(self) -> None:
        pass


class B():
    def __init__(self, a: A) -> None:
        pass

m = B(A())




